cloudify_cli/commands/apply.py

```diff
@@ -0,0 +1,188 @@
+          skip_plugins_validation,
+          parameters,
+          allow_custom_parameters,
+          logger,
+          client
+          ):
+    """Apply command uses cfy install or deployment update depends on
+    existence of DEPLOYMENT_ID deployment.
+
+    If the deployment exists, the deployment will be updated with the
+    given blueprint.
+    otherwise the blueprint will installed (the deployment name will be
+    DEPLOYMENT_ID).
+    In both cases the blueprint is being uploaded to the manager.
+
+    `BLUEPRINT_PATH` can be a:
+        - local blueprint yaml file
+        - blueprint archive
+        - url to a blueprint archive
+        - github repo (`organization/blueprint_repo[:tag/branch]`)
+
+    Supported archive types are: zip, tar, tar.gz and tar.bz2
+
+    `DEPLOYMENT_ID` is the deployment's id to install/update.
+    """
+    # check if deployment exists
+    if deployment_id not in [deployment.id for deployment in
+                             client.deployments.list()]:
+        ctx.invoke(
+            install.manager,
+            blueprint_path=blueprint_path,
+            blueprint_id=blueprint_id,
+            blueprint_filename=blueprint_filename,
+            validate=validate,
+            deployment_id=deployment_id,
+            inputs=inputs,
+            workflow_id=workflow_id,
+            force=force,
+            visibility=visibility,
+            tenant_name=tenant_name,
+            skip_plugins_validation=skip_plugins_validation,
+            parameters=parameters,
+            allow_custom_parameters=allow_custom_parameters,
+            include_logs=include_logs,
+            json_output=json_output,
+        )
+    else:
+        # Blueprint upload and deployment update
+        logger.info("Deployment {id} found, updating deployment.".format(
+            id=deployment_id))
+        update_bp_name = blueprint_id or deployment_id + '-' + datetime.now(
+        ).strftime("%d-%m-%Y-%H-%M-%S")
+        processed_blueprint_path, blueprint_id = get_blueprint_path_and_id(
+            blueprint_path,
+            blueprint_filename,
+            update_bp_name
+        )
+
+        try:
+            ctx.invoke(
+                blueprints.upload,
+                blueprint_path=processed_blueprint_path,
+                blueprint_id=blueprint_id,
+                blueprint_filename=blueprint_filename,
+                validate=validate,
+                visibility=visibility,
+                tenant_name=tenant_name
+            )
+
+        finally:
+            # When an archive file is passed, it's extracted to a temporary
+            # directory to get the blueprint file. Once the blueprint has been
+            # uploaded, the temporary directory needs to be cleaned up.
+            if processed_blueprint_path != blueprint_path:
+                temp_directory = os.path.dirname(
+                    os.path.dirname(processed_blueprint_path)
+                )
```

Comment:
double dirname for sure?

Reply:
If its url so its copied to /tmp/tmpdir/...
this is also used in cfy install .